### PR TITLE
Add go version to gopls cache key

### LINKED ISSUE #8071: `gopls` caching error when updating go version
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

First of all, thank you very much for this awesome IDE! I am just getting started using Zed but I already love it!

As discussed with @mrnugget in #7902, I opened this Issue.

I installed the new go version (`go1.22.0`), and I ran `go install golang.org/x/tools/gopls@latest` to build `gopls` with the newest go version on my system. In VSCode, everything works and `gopls` supports the new language features, but not Zed. Zed installed `gopls` when I had an older version of go installed and cached it in `~/Library/Application\ Support/Zed/languages/gopls/gopls_0.14.2`. 

`0.14.2` is the newest version of `gopls`, but it was not installed using the newest version of go and, therefore, does not support the newest language features (in my case, it was iterating over integers `for i := range 10`). If you install a new version of go, you must run `go install golang.org/x/tools/gopls@latest` again using the new version, which Zed didn't.

Deleting `~/Library/Application\ Support/Zed/languages/gopls/gopls_0.14.2` forces Zed to reinstall `gopls` using the new version of go and everything works.

One solution would be to cache `gopls` differently: Instead of naming the cached file `gopls_{GOPLS_VERSION}` (see [here](https://github.com/zed-industries/zed/blob/main/crates/zed/src/languages/go.rs#L99) or [here](https://github.com/zed-industries/zed/blob/main/crates/zed/src/languages/go.rs#L151)), we could cache it like this: `gopls_{GOPLS_VERSION}_{GO_VERSION}`. By doing that, we also cache with which go version `gopls` was built. Changing/Updating the go version then triggers a new `gopls` installation.

Checking if a globally available version of `gopls` is available would also be a nice idea and is somewhat related to this issue (See #7902 for more details).

### Environment

Zed: v0.122.2 (Zed)
OS: macOS 14.2.1
Memory: 18 GiB
Architecture: aarch64

### If applicable, add mockups / screenshots to help explain present your vision of the feature

_No response_

### If applicable, attach your `~/Library/Logs/Zed/Zed.log` file to this issue.
If you only need the most recent lines, you can run the `zed: open log` command palette action to see the last 1000.

_No response_

### PR DESCRIPTION
**Pull request by nilskch**
Closes #8071

Release Notes:

- N/A

Previously we cached gopls (the go language server) as a file called `gopls_{GOPLS_VERSION}`. The go version that gopls was built with is crucial, so we need to cache the go version as well.

It's actually super interesting and very clever; gopls uses go to parse the AST and do all the analyzation etc. Go exposes its internals in its standard lib (`go/parser`, `go/types`, ...), which gopls uses to analyze the user code. So if there is a new go release that contains new syntax/features/etc. (the libraries `go/parser`, `go/types`, ... change), we can rebuild the same version of `gopls` with the new version of go (with the updated `go/xxx` libraries) to support the new language features.

We had some issues around that (e.g., range over integers introduced in go1.22, or custom iterators in go1.23) where we never updated gopls, because we were on the latest gopls version, but built with an old go version. 

After this PR gopls will be cached under the name `gopls_{GOPLS_VERSION}_go_{GO_VERSION}`.

Most users do not see this issue anymore, because after https://github.com/zed-industries/zed/pull/8188 we first check if we can find gopls in the PATH before downloading and caching gopls, but the issue still exists.

### CODE REVIEW

**Comment by nilskch** [comment]
@mrnugget could you please take a look at this? I promise I tested this properly and it solves the issue! 

PS: I updated the PR description to provide more details on this issue

File: crates/languages/src/go.rs [hunk 1]
```diff
@@ -111,11 +112,17 @@ impl super::LspAdapter for GoLspAdapter {
         container_dir: PathBuf,
         delegate: &dyn LspAdapterDelegate,
     ) -> Result<LanguageServerBinary> {
+        let go = delegate.which("go".as_ref()).await.unwrap_or("go".into());
+        let go_version_output = util::command::new_smol_command(&go)
+            .args(["version"])
+            .output()
+            .await?;
```
Comment (mrnugget) [inline crates/languages/src/go.rs:119]: Can you add a `.context` here to make it clear that "failed to run `go version`" or something?

**Review by mrnugget — APPROVED** [review]
Looks good to me! But a small `.context` would help a lot.

(Sorry for not replying earlier! Been super busy with other things)

**Comment by nilskch** [comment]
No problem, this issue is not the biggest priority! I just try to clean up the issues I file/encounter

**Comment by nilskch** [comment]
More context, if we do not have go installed we already fail here:
https://github.com/zed-industries/zed/blob/4564273322dfbc34f20232474fe40c9988d84aaa/crates/languages/src/go.rs#L81-L106

**Comment by mrnugget** [comment]
Thank you!

**Pull request merged**